protocol/jsonlines/decoder.go

```diff
@@ -15,6 +14,7 @@ const (
 
 type lineReader interface {
 	ReadLine() ([]byte, bool, error)
+	ReadSlice(delim byte) (line[] byte, err error)
```

Comment:
Mmm, I think the code is simpler this way; [ReadSlice] https://golang.org/pkg/bufio/#Reader.ReadSlice returns a pointer (or pointers since it's a slice) to the original slice and don't block so it's perfect for the read-discard done on discardPending (no copying and no blocking). ReadLine on the other hand does a copy but also returns a specific code if the buffer is full so we can continue iterating by chunks.

Both could be implemented using the other but I think it's simpler that way in the places they're used here.